### ISSUE TITLE
Python Recipe (abstract) base class.

### ISSUE THREAD

**Issue by brenthuisman** [arbor-sim/arbor#1634]
It would be nice to have `arbor.recipe` do a few more things for the user:

* Throw error when required methods are not overridden.
  * `PYBIND11_OVERRIDE_PURE` is not throwing in the current `arbor.recipe` impl. Only upon use of the instantiated class derived from `arbor.recipe`. Not entirely sure why, but earlier surely would be nice.
* Do away with manually having to call `__init__` on the `arbor.recipe` base class.
* Maybe: have `print()` do something useful.
* Question: does `arbor.recipe` need to be a abstract base class [2]?

See [1] for the trampoline pattern.

[1] https://pybind11.readthedocs.io/en/stable/advanced/classes.html
[2] https://docs.python.org/3/library/abc.html

**Comment by bcumming** [comment]
We are seeing the limits of trying to directly wrap the C++ ABC recipe implementation in Python, and it might be worth considering taking a different approach from the Python side that is less surprising to Python users.

As an alternative to the current recipe interface: use a dictionary.
```python
# example stand alone function that implements part of recipe interface
def cell_count_foo(state):
    return state["cell-count"]

# state can be any arbitrary data the user wants to
my_state = {"cell-count", 42}

# classes are just fancy dictionaries anyway...
my_recipe = {"cell_kind": get_cell_fun,
             "cell_description": description_fun,
             "cell_count": cell_count_foo,
             "state", my_state} # required: passed to each of the recipe interface functions

domdec = partition_load_balance(my_recipe, context)
```

As an added bonus, calls like `partition_load_balance` could test whether the parts of the recipe interface that they need are provided. That way we could implement my suggested fix for Robin's connection ordering question. You wouldn't need to implement all of a recipe interface, if it isn't required by the api method you pass it to.

**Comment by brenthuisman** [comment]
Another option is to have Pybind expose the `recipe` base class privately (start w/ underscores), wrap it in a Python class (as in, written in Python in `__init__py`) and then use the `@abstractmethod` decorator to mark some methods as required to have implementation.

I don't know of a way around calling the base class constructor, afaik that's always a necessity in Python.

**Comment by thorstenhater** [comment]
Essentially classes in Python have a dictionary of their methods (`obj.__dir__()`).  The proposed approach
 via a `dict` seems to be coming close to re-implementing the Python object model and virtual inheritance.
So, the functionality to test for a method is already there, eg
```py
if hasattr(D, foo) and callable(getattr(D, foo)):
    print("OK")
```

About the base class constructor problem
```py
class B:
    def foo(self):
        print("B::foo")

class D(B):
    pass

d = D()
d.foo()
```
This works just fine. However, if you have a custom `__init__` method that sets values and you want
those in the derived class, then you need to call `super().__init__(args)`, but this is normal in Python.
Thus, I say it is fine to expect it from our users, since they are usually familiar with Python.

The `@abstractmethod` is not really doing much, in principle it will just `raise AbstractMethod(method_name)`.
 If we want to refrain from having more Python code in the code base (for now!) we can just emulate
 this by raising similar exceptions manually. By extension using `abc` can be similarly avoided by hooking into
 the metaclass of the object to `__subclasshook__` derived classes' method definitions and checking whether
 they are abstract.
 
 I think the core issue here is that the virtual inheritance as built by our pybind11 approach does not do
 what we expected it to and the proper fix is to review and address that. Proposal:

- Do not 'fix' the requirement to call `super().__init__(...)`  in derived recipes; it's proper Python OO.
- Fix the inheritance by forcing overrides of conceptually abstract methods
    - Note that we told some of our users to exploit that we do not need all methods for `partition_load_balance`
       to optimise two-phase init of connection tables. We might want to make this (more) explicit.

**Comment by akuesters** [comment]
I agree with Thorsten not to "fix" the `__init__` call. Forcing overwrites should be fixed though.

**Comment by thorstenhater** [comment]
This seems to be a known issue w/ pybind11, ref https://github.com/pybind/pybind11/issues/1978

**Comment by brenthuisman** [comment]
The actual request is I think to have feedback on whether or not the recipe is complete at time of recipe instantiation, not `partition_load_balance()`, which may happen on a different system in a faraway land.

@bcumming solution (`recipe_dict`?) is one. Perhaps a way to test a recipe is another. We also could document the pattern of instantiating a domain decomposition (that you do not necessarily use) as a way of testing.

@Helveg do you have a preference?

**Comment by Helveg** [comment]
A class is preferred over a dictionary for UX. You'd have to define functions, then reuse the same names to put them into a dict:

```
class recipe:
  def B(): pass

# vs

def B(): pass

recipe = dict(B=B)
```

A class is also prettier in syntax highlighting :sunglasses:.

The forced init call for C++ memory reasons seems strange. Can't `__new__` create a correct new object? `__new__` is for creating new objects, `__init__` is solely for initializing their attributes, `__new__` is also much less frequently used by users and can be expected to break things when you mess around with it. Having an object that segfaults or something if you dont init is not a Python thing and not expected to be due to inheritance, so I'd say it's confusing.

I can't say much about the best way to work around the pybind stuff

@thorstenhater few nitpicks that might interest you: `__dir__` is used only for when users want to inspect an object with `dir()` and returns a list of attributes to be played with. It isn't involved in attribute lookup afaik. It is `__dict__` that is a class's `mappingproxy` and instance's attribute dictionary. `__init_subclass__` might also be a better fit than `__subclasshook__`.

**Comment by thorstenhater** [comment]
@Helveg Yes, using `__dict__` is the correct way.

I am less sure about `__new__` though, does it automatically call `super().__new__()`? If not,
we might have the same issue under a different name.

------------------------------------

@brenthuisman 
> The actual request is I think to have feedback on whether or not the recipe is complete at time of recipe instantiation, not `partition_load_balance()`, which may happen on a different system in a faraway land.

The original issue comments on the fact that abstract methods defined via `pybind11` are not 
abstract. 

> @bcumming solution (`recipe_dict`?) is one. Perhaps a way to test a recipe is another. We also could document the pattern of instantiating a domain decomposition (that you do not necessarily use) as a way of testing.

This solution is equivalent (in Python!) to having an object to which methods are added on 
later. It's the correct thing to do (see below), it's just that this is going against the grain of the language.
I pointed out that we can have the best of both worlds: using OO common to Python folks 
and having `XYZ_load_balance` to check for required methods.

---------------------

Here's another option: lean into monkey patching even more
```py
rec = arb.make_recipe() # return a corretly initialised recipe object, not a class
# Successively add/overload members
rec.cell_kind = lambda self: return arb.cable_cell
```
or we can make a builder pattern
```py
rec = arb.make_recipe() \
               .cell_kind(lambda self: return arb.cable_cell) \
               ...
```
and allow checking if this object is suitable for use in `XYZ_load_balance`
at any step. 
@Helveg what do you think about this idea?

Reasoning here is that recipes are essentially classes with one instance for most
Arbor simulations. This is why I like @bcumming's idea, but not the implementation.
For those case where this is not true, eg parametrised recipes, the old OO way 
would still be accessible.

**Comment by Helveg** [comment]
I dont' see call chaining all that often in Python, probably because how line wrapping of arguments is usually done under PEP8. And adding the backslashes to have it Javascript style is a pain.

Moving it to `__new__` might be moving the goalposts, but the goal is open. EOF sports analogy, nobody overrides `__new__`, `__init__` is commonly used. Sadly the scriptures can be interpreted both ways:

> If a base class has an `__init__()` method, the derived class’s `__init__()` method, if any, must explicitly call it to ensure proper initialization of the base class part of the instance; for example: `super().__init__([args...])`.
>
> Because `__new__()` and `__init__()` work together in constructing objects (`__new__()` to create it, and `__init__()` to customize it), no non-None value may be returned by `__init__()`; doing so will cause a `TypeError` to be raised at runtime.

`__new__` should fully construct it, which to me as a Python user sounds like: All binary guts need to be A-okay, either by letting CPython's `object` or `type` construct/allocate/wazoozle them, or to let `arbor`s internals create the object, but I should be handed a functional object that I can start customizing in `__init__`. The scriptures do prescribe the `super()` so I can't really object to relying on it, but why would we want to put an unclear C++ pitfall below a commonly made mistake? I don't understand what the core issue is but let's at least put the pitfall under the much less used `__new__`.

**Comment by thorstenhater** [comment]
Hi,

I noticed this has gone stale. Can we close this?

**Comment by brenthuisman** [comment]
Pandora's box off call chaining has been opened, so that could address init. When could we be sure no further change to the object is required though? Method chaining still add things in one by one, so we must add a way to signal that more is coming, or that the recipe can be checked now, which seems not any more user friendly than just adding a `check()` method.

That and `print()` (goes for many other objects too) are still things that are not stale.

**Comment by Helveg** [comment]
Also the binary guts spilling when `super().__init__` isn't called, binary should be solved on `__new__` , there are multiple cases where object creation doesn't even call `__init__` (like pickling, as textbook example), just `__new__`

**Comment by thorstenhater** [comment]
Ok, it has been about a year since the last update.
Taking up the discussion
>    Throw error when required methods are not overridden.
        PYBIND11_OVERRIDE_PURE is not throwing in the current arbor.recipe impl. Only upon use of the instantiated class derived from arbor.recipe. Not entirely sure why, but earlier surely would be nice.
    Do away with manually having to call __init__ on the arbor.recipe base class.
    Maybe: have print() do something useful.
    Question: does arbor.recipe need to be a abstract base class [2]?

1. `print`: is covered by the text repr/object printer tickets. So, I think we should discuss that there.
2. `__init__` having to be called is part of Python's object model. Dislike it as much as you want, but there's no way around it. 
3. Which methods _must_ be overriden? There's some combinations that go together, but no single method, right?

Looking at it, maybe split out 2. and 3. into different issues and closing this might help the flow of discussion.

**Comment by Helveg** [comment]
2. That's not true.

**Comment by thorstenhater** [comment]
The base class' `__init__` is not called automatically in Python in contrast to C++ and other OOy languages.
So, if we want to call the base's initialisation, we have to do that by hand.

https://stackoverflow.com/questions/1385759/should-init-call-the-parent-classs-init

**Comment by Helveg** [comment]
Yes, but not calling `__init__` is a valid possibility that is now blocked by a mostly uncaught error. Just move whatever is in `__init__` that doesn't need arguments to `__new__`, then let `__init__` set some values to initiliaze the object. That's following the proper Python object protocol. `__new__` for creation of the `PyObject` of the correct type, `__init__` for setting values coming from the constructor (and not calling `__init__` being a valid choice for people venturing out on their own. You don't need to endorse their silliness, but shouldn't block them either).

**Comment by thorstenhater** [comment]
That case is _fine_, since we did never override `base.__init__` so `derived()` will call that.

**Comment by Helveg** [comment]
It's not fine, during valid object creation, it is possible that `__init__` is deliberately never called at all.

**Comment by thorstenhater** [comment]
Yes, but to get a properly set up object, they still need to call `__init__`

**Comment by brenthuisman** [comment]
I could not definitively establish whether or not `super().__new__` is auto-called in Python. Is it?

Also, working around this through `__new__`, which is not how `pybind11` works, is also messy. We're going to mix them up on purpose in order to achieve a slight convenience.

**Comment by Helveg** [comment]
Yea, I'm not familiar with `pybind11`, but it seems key to `pybind`s default flow to do things through `__init__`, however, there is actually no problem because:

> Changed in version 2.6: The default pybind11 metaclass will throw a TypeError when it detects that init was not called by a derived class.

```
>>> class X(arbor.recipe):
>>>  def __init__(self): pass
...
>>> X()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: arbor._arbor.recipe.__init__() must be called when overriding __init__
```

There is no messiness towards the user, it's clear what is expected, and indeed, no need to go against the flow here. So I considered the debate settled towards leaving `__init__` as is.

**Comment by brenthuisman** [comment]
We weren't (I checked). It was worth exploring whether or not it could be automated (factory is still an answer), but nothing stands out as 'good' to me.

Ensuring proper recipe state, OTOH, that's then the last thing we must decide before closing: what's a good way to make sure people's `recipe`s are complete?

**Comment by thorstenhater** [comment]
Yeah, that has been my opinion as well for some time.

Question @brenthuisman: How do you define 'complete'?
```py
class R(arb.recipe):
  pass
```
is complete, although not helpful. 
```py
class R(arb.recipe):
  def numcells(self):
     return 42
```
is obviously incomplete, but what should we flag?

Even worse, since we are in Python land, anyone can come at any time and patch up the 'missing' methods.

**Comment by brenthuisman** [comment]
'Complete' can be defined as 'has implemented required overloads' at minimum.

**Comment by thorstenhater** [comment]
Ok, then define 'required overloads'

**Comment by thorstenhater** [comment]
Update: metaclasses and pybind11 are not friends. At all.

**Comment by brenthuisman** [comment]
We cannot solve it the preferred way, but we can address it by having `arbor.simulation` check. Acceptable?

**Comment by brenthuisman** [comment]
0. You could call them and watch for errors, no?
1. There could be a Python recipe validator function (check if a user defined recipe has the right methods). This would let users be able to test if they created a proper recipe.
2. Pybind already throws up if you don't implement the virtual functions when instantiating `arbor.simulation`. E.g. `RuntimeError: Tried to call pure virtual function "py_recipe::cell_kind"`. We could catch these and translate to its Python alias.


Also, did you know this is the shortest valid recipe?

```python
class r (arbor.recipe):
    def num_cells(self):
        return 0
```

**Comment by thorstenhater** [comment]
0. We are doing this when constructing the simulation.
1. As you noted above, just `num_cells` is enough. So, not much to gain.

About translating the error: There is no dedicated error type in Python (not even in ABC, seemingly?!)
for this situation. We wouldn't be helping users much by adding one.

As always, I am really interested in an improvement here, but I absolutely don't see a way.

**Comment by brenthuisman** [comment]
`num_cells` isn't enough, if it doesn't return 0 :) So, it actually does seem to do a full check already. It's just that it happens a bit late (when passing the recipe to `simulation`) and gives Pybind C++ function names, not their Python bindings.

There's no type, but there's a pretty unique string (e.g. it includes `py_recipe::cell_kind`). We could match against that.

**Comment by thorstenhater** [comment]
Regarding the time:
```py
class recipe(A.recipe):
  pass

sim = A.simulation(recipe())
```
where would you do the check (earlier)?

> Pybind C++ function names, not their Python bindings.

Could you provide an example expected ./. reality?

> There's no type, but there's a pretty unique string (e.g. it includes py_recipe::cell_kind). We could match against that.

I do not understand that comment, however, in general I'd stay as far as possible from matching on strings.

**Comment by brenthuisman** [comment]
See above for the example. Repeat:

```python
class r (arbor.recipe):
    def num_cells(self):
        return 1
sim = arbor.simulation(r())
```
will print `RuntimeError: Tried to call pure virtual function "py_recipe::cell_kind"`. And so on. `s/py_recipe::/recipe./g`?

**Comment by thorstenhater** [comment]
Ok, that resolves one of three issues you quoted. Could we then close this and make one specific issue for throwing a better exception?
